orchestrator/clickhouse/geoip/database.go

```diff
@@ -36,13 +53,23 @@ func (c *Component) openDatabase(which string, path string, container *atomic.Po
 	if err != nil {
 		return err
 	}
-	oldOne := container.Swap(&newOne)
+	c.db.lock.Lock()
+	var oldOne geoDatabase
+	switch which {
+	case "asn":
+		oldOne = c.db.asn[path]
+		c.db.asn[path] = newOne
+	case "geo":
+		oldOne = c.db.geo[path]
+		c.db.geo[path] = newOne
+	}
+	c.db.lock.Unlock()
 	c.metrics.databaseRefresh.WithLabelValues(which).Inc()
 	if oldOne != nil {
 		c.r.Debug().
 			Str("database", path).
 			Msgf("closing previous %s database", which)
-		(*oldOne).Close()
+		oldOne.Close()
```

Comment:
I think there is a race condition. You need to keep the write lock up to closing...